main.cc

```diff
@@ -380,10 +380,15 @@ class Peer : public SocketHandler {
 
 		// Some error on the socket.
 		state_ = STATE_TCP_CONNECTING;
-		if (errno != EINPROGRESS && errno != EAGAIN) {
-			if (!stat.tcp_connections)
-				throw Except("cannot establish even one"
-					     " TCP connection");
+		if ((errno != EINPROGRESS && errno != EAGAIN) || ret) {
```

Comment:
Due to the condition above, `ret` is always true here, so the change just disables `add_to_poll()` at the below - why? We come here in 2 cases: the system call has been interrupted by a signal or it's still in progress, so why to remove the socket from the `epoll()` queue?

Reply:
Ah, I see. Yes, badly fixed. But there still a bug. Let ne explain reason for the fix.

There is also another case: connection can be never established, e.g. server is not started (and connection is refused) or no route to host is found. Or server has crashed right in the middle of test...

When a nonblocking `connect()` fails, the following happen:
```
return code = -1
errno = EINPROGRESS
SO_ERROR = ECONNREFUSED or EHOSTUNREACH or smth else
```

On current version when we try to reach unreachable server `connect()` fails and sets variables as listed above.

In `tcp_connect_try_finish()` we first read `SO_ERROR` and save that value to `ret` variable. Note, that `SO_ERROR` is reset to `0` after we read it. Since `ret` is non-zero, we do some error processing.

Since `errno == EINPROGRESS` we just insert the socket back into the epoll.

On the next `epoll_wait()` the socket is returned as ready to write and we again get into `tcp_connect_try_finish()` . But this time variables has different values:
```
ret = -1
errno = EINPROGRESS
SO_ERROR = 0 
```

First we read `SO_ERROR` and add save it to `ret`. `ret` is `0` now and we try to perform a TLS handshake over connection which wasn't open. And a crash happen somewhere inside openssl.

This happens because we ignore error from `SO_ERROR` in that condition on line 383. 
Finally, we need to log not `errno` value, but `SO_ERROR`

Reply:
Yeah, I see. I'd say that there is no sense to keep after-`connect()` and after-`epoll()` workflows in the same `tcp_connect_try_finish()`. This code is mainly left from the original THC tool. It'd be much more straightforward to call `tcp_connect_try_finish()` after `epoll()` only and move `connect()` specific code to `tcp_connect()`. Also in this case we won't do the `getsockopt()` call right after `connect()`, i.e. save one system call and that's good.